### ISSUE TITLE
deriveAndLocal can't replace/transform a setting that's already defined

### ISSUE THREAD

**Issue by dansanduleac** [sbt/sbt#1034]
On the last line of `deriveAndLocal`, derived settings are _prepended_ to the other settings. There's a comment too shortly before, stating this: 

```
// valid derived settings to be added before normal settings
```

That means it's impossible to transform an existing setting, which seems like something one might want to do, because it would just get replaced by the "original" later on.
Is there a specific reason for this approach? If not, I'd suggest appending the derived settings instead.

**Comment by harrah** [comment]
Yes, there is a reason.  Individually, explicitly defined settings should override automatic settings.

**Comment by dansanduleac** [comment]
That makes sense. However, it may be useful to allow derived settings to come after `Defaults.defaultSettings`. Then, one would be able to conditionally transform these settings if some other settings are defined. 
If that's not possible, maybe I'm looking for another feature: a derived setting that is only set if its prerequisites have already been defined strictly _before_ it (so without doing any DAG reordering).

**Comment by harrah** [comment]
It would be good to provide a use case.
- Automatic settings that the user cannot override have been problematic in practice.
- Giving derived settings an ordering in relation to normal settings is incompatible with their basic function.  If a user explicitly overrides a leaf setting, it is supposed to derive the intermediate setting.  But, the user override would come later in the settings sequence and so it wouldn't trigger the derived setting.
- Related, a setting definition does not currently care about the order of other settings for keys it uses.  It is only for key it is defining.  This would change that.
- So, what happens here:

``` scala
a := 3
newDerive(b := a.value)
a := 5
```

The proposal is that b is injected only because of `a := 3` (it doesn't see `a := 5`), but the normal model is that `b` sees `5` as the value of `a`.  So, there is an inconsistency between what triggers derivation and what is used for the value.

**Comment by dansanduleac** [comment]
My use case is that I have some settings, and I want either _all of them_, or _none of them_, to be defined (or transformed) conditionally on whether some other set of settings is defined (in the previously mentioned case, these "triggers" are `team`, `kind` and `visibility`). I suspect it could more clearly be solved with the idea of "[features](https://github.com/sbt/sbt/pull/998#issuecomment-29137666)" that you mentioned back when we were discussing #998. I have pretty much the same use case as back then.

The specific issue I've now encountered is in conditionally transforming `projectId`. I tried to do (slightly simplified):

``` scala
Def.derive(projectID <<= (projectID, someTrigger) { case (pid, _) =>
  pid extra ("someAttr" -> "a value")
})
```

But even though it does get derived (I checked), it gets completely overridden by its definition in `Defaults`: `projectID <<= defaultProjectID`. Therefore there's no mention of my definition when I ultimately `inspect projectId`.
Essentially, what I think I want is **for each `Setting` derived from a `DerivedSetting` to end up in the same position as the `DerivedSetting`**. 
That would mean 
- injecting the deriveds in place of the `DerivedSetting` for each `DerivedSetting` in the settings seq
- Making `DerivedSetting` _not_ inherit `DefaultSetting` anymore

I agree with you that derived settings that cannot be overridden by the user are bad. I'm not proposing that. In this case the user can just override it _after_ declaring the `DerivedSetting` and all's good. The setting would get derived, and then the derived value would get overridden. One issue I can see is that the setting _might_ get derived in more specific scopes than the scope where the user overrides it, but then the derived setting would prevail anyway for those scopes, regardless of where it ends up in the settings seq.

Re: your example, let me see if I get what you mean:
- first bullet point - addressed above
- Not sure what you mean by "it is supposed to derive the intermediate setting" - who is?
- I'm not saying that the derived settings should "only look to the left" for triggers, or anything like that, and then be defined according to those values, and not according to the later-overridden values of the dependencies. As far as I'm aware that's not happening now and not sure it would be easy to do either. At the point of derivation there is no mention of the `.init`s. Therefore as far as I can tell, right now (and with this proposed change as well), "there is an inconsistency between what triggers derivation and what is used for the value."
- In your example I would expect it to work the same as it does now (a = b = 5). However, in this example:

``` scala
a := 3
b := 6
newDerive(b := a.value)
a := 5
b := 8
```

I'd expect `a=5, b=8`, but if I remove the `b := 8` at the end, I'd expect `a=5, b=5`.
As far as I can tell `deriveAndLocal` doesn't differentiate between `a := 3` and `a := 5`, they both just mean "a is being defined so can therefore trigger things".

**Comment by harrah** [comment]
Ok, I think you have a good argument there.  You want them to be injected where they are defined: not appended or otherwise change the model.  What are the obstacles you see to implementing this?  Is the current implementation amenable to this?  In your third point, did you mean to disagree with "there is an inconsistency"?  It seems to me after understanding your clarified argument that there is NOT an inconsistency.  Dropping DerivedSetting from being a DefaultSetting seems ok off hand.  The main related change is probably to make existing uses of derived in sbt be default settings explicitly.

**Comment by dansanduleac** [comment]
Yes, exactly! 
I've already whipped up an implementation and it seems to work for my use case. I ran a quick test and see some tests failing, but then again I'm on 0.14 so not sure if it's because of my change or not, will look into it.
Hmm that's odd, I meant to agree... but in the idea that "the _values_ of the defined deps that a setting gets derived from are not necessarily the values that the derived will init from, because the dep's _latest_ values will be used". Not that it really matters. I mean that was always the case anyway.
I'll start a pull request with what I have now that I think works, so you can have a quick look if you want, and I will change the existing uses to be default settings explicitly a bit later, probably tomorrow as I'm about to head off.

**Comment by harrah** [comment]
Ok, sounds good.  To be clear, this will still work, right?

``` scala
newDerive(b := a.value)
a := 3
```

**Comment by dansanduleac** [comment]
Yes

**Comment by dansanduleac** [comment]
While merrily using this new implementation that I'm proposing, I've stumbled upon another issue (or at least I think it is):
The derived settings are always derived in the scopes of one of the dependencies (namely, for each scope where some setting gets derived, that scope will be taken from whichever dependency 1) is a trigger and 2) is processed last, after all other dependencies are/have already been defined in its scope).
However, the scope where the user places the `DerivedSetting` is ignored.

For instance, say I'm redefining `projectId` as before, in a plugin, somewhere in `projectSettings` (anywhere else like buildSettings / globalSettings, it gets replaced by the project default anyway).
I go 

``` scala
override def projectSettings = Seq(
  Def.derive(projectID <<= (projectID, someTrigger) { case (pid, _) =>
    pid extra ("someAttr" -> "a value")
  }, trigger=(_ == someTrigger.key))
)
```

Now if I have 2+ (`n`) projects, each one will contribute a `DerivedSetting(projectID <<= ...)` scoped in its project scope to the build's `Seq[Setting[_]]`. If then this trigger is defined in each project (so also n times), what will happen is that it will cause _both_ (or all `n`) of those `DerivedSetting`s to happily derive 1 setting per project. So we get something like n*n settings in total.
All of this isn't apparent until you derive a setting that does a `<+=` or such, and then you end up with duplicate identical settings added to that collection, for each project.

I'm proposing we fix this by taking into account the _scope_ where the `DerivedSetting` was defined by the user (i.e. its `(setting.key:ScopedKey).scope`), and only allowing the derivation _under_ that scope.
Specifically, let's say we have a derived setting `derived`, and
- `derivedScope` is the scope where my `derived` was defined (so if `derived` was added to a plugin's projectSettings, then this would be the project's scope, for a given project), and 
- `depScope` is the scope of a triggering dep of `derived`
  then
- if they **intersect** (one includes and is more specific than another, with regard to delegation), then derive the actual setting using the more specific scope of the two
- otherwise, don't derive

This way, even if all the dependencies have been defined in Global, let's say, my derived setting wouldn't be derived under Global, but under each project's scope (still assuming it was added to a plugin's projectSettings). This seems more intuitive to me because it doesn't discard the scope of the derived setting.

NB: I thought about achieving the same with the _filter_ param of `derive()` but unfortunately I can't say "this project". I've tried `fold`ing the scope's ProjectAxis, but even if I provide an `ifThis` parameter of `true`, it seems it never gets derived. And of course I can't directly use the value of `thisProjectRef`.

So, thoughts?

**Comment by harrah** [comment]
- Do you really mean intersect?  I would think you'd only want if `depScope` delegates to `derivedScope` and not the other way.
- `This` and `ProjectReference`s besides `ProjectRef` and `BuildRef` get resolved before the `Seq[Setting[_]]` is passed off to the settings system, which is why you don't see `This`.
- Derived indeed wasn't designed with the expectation of using `+=`.
- I think this is a fine enhancement- it doesn't affect existing uses because existing uses are all in `GlobalScope`, right?
- As a general comment, delegation is relatively expensive in terms of setting processing time.  For example, looking up `aggregate` in every scope (to build the aggregation index) takes 25% of indexing time (~0.5 s for the sbt build), largely because the lookup ultimately delegates to the global scope and so there are lots of map lookups along the way.  In this case, they aren't lookups, but they are still equality checks.  If it becomes a problem, one optimization is to short-circuit when `derivedScope` is `GlobalScope`.

**Comment by dansanduleac** [comment]
- The reason why I also accept when `derivedScope` delegates to `depScope` is that my dependencies might all be in `ThisBuild`, for instance, but my derived setting was defined in `projectSettings`. In this case, I can do two things (not sure which of them you were suggesting):
  - **not derive**; however, this behaviour would be inconsistent with the case where this setting would be a normal setting instead. The reason why I make this comparison is because I view a derived setting as a sort of a superset of a normal setting - if a normal setting can be defined successfully in some scope (since all the dependencies are available), then I think that a DerivedSetting should also be derived in that scope (if not in others as well). 
  - **derive**: but where? Deriving in the scope of the dependency, the build scope, would now mean that we get (# of projects) many settings in ThisBuild, which would either override each other (assignment) or chain-accumulate (`+=` and friends). Clearly not what we want.  Deriving a setting in each project's scope is what one would expect, and it's also how a normal setting would behave.
- It doesn't affect existing uses (compared to 0.13.0), but the existing uses are actually not in `GlobalScope`.. they are in fact in `Scope(This, This, This, This)` as they aren't scoped by default. That means that they (only 3: `scalaBinaryVersion`, `testGrouping` and `compilersSetting`) end up scoped under each project's scope, one setting per project (or in the case of `testGrouping`, for some reason, 3 identical settings per project). Not sure if that's the intended behaviour.
- Fair enough, I'll put that optimisation in place and update my pull request with these changes.

**Comment by dansanduleac** [comment]
Actually I found the answer myself for why there are more than 1 of those settings: back when they were `DefaultSetting`s they would get collapsed and only the first occurrence kept (if the comment above `class DefaultSetting` is to be trusted).

**Comment by dansanduleac** [comment]
Fixed by #1036
